### PR TITLE
Add voting procedures in DAO

### LINKED ISSUE #228: DAO should be able to raise funds through a Voting process
The goal is to be able to restart the DAO funding through a Voting, much like when proposing a new venture, or releasing dividends. 

However, that would require the DAO (or the Voting?) to call on the DAO itself several methods from `IssuanceEth` (for example, `setIssuePrice`, or `startDistribution` etc.) which can be called only by the owner (the deployer of the DAO, that is). 

This task is also to find a possible solution to this dilemma.

One way of bypassing the issue would be to have the owner call on the DAO `transferOwnership` to the DAO once the DAO is launched.

### LINKED ISSUE #239: Implement a generic `propose` function in DAO.sol.
We could allow the DAO to accept proposals with arbitrary proposalData, so that a proposal can be done for any single function and parameter combination in the DAO contract. Instead of proposeVenture and proposeDividends we would have propose(contractAddress, proposalData)

There are some major setbacks: 
1. It would require participants to manually check (if it's available) the contract code of every proposal in order to know what they are voting for. 
2. It would also make room for memory abuse of the proposals storage variable.
3. We still have to keep the individuated enacting functions in the DAO.

Discuss.

### LINKED ISSUE #243: DAO.sol needs a proposal to cancel an investment.
Otherwise failed issuances will keep the DAO ether locked in.

### PR DESCRIPTION
**Pull request by uivlis**
Fixes #228, fixes #239, fixes #243.

### CODE REVIEW

File: contracts/examples/dao/DAO.sol [hunk 1]
```diff
@@ -19,13 +19,11 @@ import "./../../voting/Voting.sol";
...
+contract DAO is VentureEth, Democratic {
 
     using SafeMath for uint256;
     using EnumerableSet for EnumerableSet.AddressSet;
 
-    event VentureProposed(address proposal);
-    event DividendsProposed(address proposal);
     event VentureAdded(address venture);
 
     uint256 public threshold;
```
Comment (albertocuestacanada) [inline contracts/examples/dao/DAO.sol]: We don't need to declare `proposals`, it's now part of `Democratic.sol`.

File: contracts/examples/dao/DAO.sol [hunk 2]
```diff
@@ -54,28 +56,20 @@ contract DAO is VentureEth {
...
-        Voting voting = new Voting(
-            address(this),
-            address(this),
-            abi.encodeWithSignature("investVenture(address,uint256)", venture, investment),
-            threshold);
-        proposals.add(address(voting));
-        emit VentureProposed(address(voting));
+    function startDistribution() public onlyOwner {
+        // solium-disable-next-line security/no-low-level-calls
+        (bool success, ) = address(this).delegatecall(
```
Comment (albertocuestacanada) [inline contracts/examples/dao/DAO.sol]: Please remind me again why can't we just call `transferOwnership(address(this))`

File: contracts/examples/dao/DAO.sol [hunk 3]
```diff
@@ -119,47 +118,44 @@ contract DAO is VentureEth {
...
+        _transition("SETUP");
+        this.setIssuePrice(_issuePrice);
+        this.startIssuance();
     }
 
-    /** Enumerators */
+    /**
+     * @notice Hook for proposals to start distribution in a non-initial investment round.
+     */
+    function restartDistribution() public onlyProposal {
```
Comment (albertocuestacanada) [inline contracts/examples/dao/DAO.sol]: Maybe we could do an `if` in `startDistribution` instead.
`if (owner() != address(this)) transferOwnership(address(this))`

File: contracts/examples/dao/DAO.sol [hunk 4]
```diff
@@ -119,47 +118,44 @@ contract DAO is VentureEth {
...
+        _transition("LIVE");
+    }
 
     /**
-     * @notice Returns the invested ventures.
+     * @notice Hook for proposals to cancel all new investments in a non-initial investment round.
      */
-    function enumerateVentures() public view returns (address[] memory) {
-        return ventures.enumerate();
+    function cancelAllNewInvestments() public onlyProposal {
```
Comment (albertocuestacanada) [inline contracts/examples/dao/DAO.sol]: Maybe we can call this `cancelInvestmentRound`. Just aesthetics.

File: contracts/examples/dao/DAO.sol [hunk 5]
```diff
@@ -54,28 +56,20 @@ contract DAO is VentureEth {
         revert("Withdraw is disabled.");
```
Comment (albertocuestacanada) [inline contracts/examples/dao/DAO.sol]: Do we still need to disable withdraw? If it is `onlyOwner` and the contract is the owner from the first distribution round, it seems to me that no one will be able to execute it.

File: contracts/examples/dao/DAO.sol (hunk 2 above)
Comment (uivlis) [reply contracts/examples/dao/DAO.sol]: Here's why: when you call `transferOwnership(address(this))`, the `msg.sender` in the transfer call becomes `address(this)`, not `owner` (which was the one who initiated the call to `startDistribution`). Therefore, you have to delegate the call, that is, call with `msg.sender = tx.origin = owner`.

File: contracts/examples/dao/DAO.sol (hunk 3 above)
Comment (uivlis) [reply contracts/examples/dao/DAO.sol]: startDistribution can be initially called only by the owner, while restartDistribution only by a voting.

File: contracts/examples/dao/DAO.sol (hunk 2 above)
Comment (albertocuestacanada) [reply contracts/examples/dao/DAO.sol]: I get it now, thanks!

File: contracts/examples/dao/DAO.sol (hunk 3 above)
Comment (albertocuestacanada) [reply contracts/examples/dao/DAO.sol]: Got it, thanks.

**Pull request merged**